ouroboros-consensus-cardano/src/shelley/Ouroboros/Consensus/Shelley/Node/DiffusionPipelining.hs

```diff
@@ -1,16 +1,117 @@
+-- pipelining.
+data HotIdentity c = HotIdentity {
+    -- | Hash of the cold key.
+    hiIssuer  :: !(SL.KeyHash SL.BlockIssuer c)
+  , -- | The issue number/opcert counter. Even if the opcert was compromised and
+    -- hence an attacker forges blocks with a specific cold identity, the owner
+    -- of the cold key can issue a new opcert with an incremented counter, and
+    -- their minted blocks will be pipelined.
+    hiIssueNo :: !Word64
+  }
+  deriving stock (Show, Eq, Ord, Generic)
+  deriving anyclass (NoThunks)
+
+data ShelleyTentativeHeaderState proto =
+    -- | Legacy state, can be removed once mainnet is in Conway.
+    LegacyShelleyTentativeHeaderState !(SelectViewTentativeState proto)
+  | ShelleyTentativeHeaderState
+      -- | The block number of the last trap tentative header.
+      !(WithOrigin BlockNo)
+      -- | The set of all hot identies of those who issued trap tentative
+      -- headers for the recorded block number.
+      !(Set (HotIdentity (ProtoCrypto proto)))
+  deriving stock (Show, Eq, Generic)
+  deriving anyclass (NoThunks)
+
+-- | This is currently a hybrid instance:
+--
+--  - For eras before Conway, this uses the logic from
+--    'SelectViewDiffusionPipelining' for backwards-compatibility.
+--
+--  - For all eras since Conway, this uses a new scheme: A header can be
+--    pipelined iff no trap header with the same block number and by the same
+--    issuer was pipelined before. See 'HotIdentity' for what exactly we use for
+--    the issuer identity.
+--
+-- Once mainnet transitioned to Conway, we can remove the pre-Conway logic here.
+instance
+  ( ShelleyCompatible proto era
+  , BlockSupportsProtocol (ShelleyBlock proto era)
+  ) => BlockSupportsDiffusionPipelining (ShelleyBlock proto era) where
+  type TentativeHeaderState (ShelleyBlock proto era) =
+    ShelleyTentativeHeaderState proto
+
+  initialTentativeHeaderState _
+    | L.eraProtVerLow @era < L.eraProtVerLow @StandardConway
+    = LegacyShelleyTentativeHeaderState NoLastInvalidSelectView
+    | otherwise
+    = ShelleyTentativeHeaderState Origin Set.empty
+
+  updateTentativeHeaderState bcfg hdr@(ShelleyHeader sph _) = \case
+      LegacyShelleyTentativeHeaderState st ->
+        LegacyShelleyTentativeHeaderState <$>
+          updateTentativeHeaderState
+            (SelectViewDiffusionPipeliningBlockConfig bcfg)
+            (SelectViewDiffusionPipeliningHeader hdr)
+            st
+
+      ShelleyTentativeHeaderState lastBlockNo badIdentities ->
+        case compare (NotOrigin (blockNo hdr)) lastBlockNo of
+          LT -> Nothing
+          EQ -> do
+            -- TODO Usually (especially during syncing), @badIdentities@ will be
+            -- empty. However, @hdrIdentity@ will still be forced by
+            -- 'Set.notMember'. According to the comment on 'isSelfIssued' for
+            -- @BlockSupportsMetrics ShelleyBlock@, we could save 850ns for
+            -- hasing the cold key in this case.
```

Comment:
hasing  -> hashing

Reply:
One microsecond seems inconsequential?

Reply:
```haskell
            -- One could even replace @badIdentities@ by a list of the unhashed
            -- cold keys when its cardinality is small, for reasons similar to
            -- the ones in the comment mentioned above.
```
For ease of reference:
https://github.com/IntersectMBO/ouroboros-consensus/blob/af4b27ed334a9a08de1be3c78e8a76f3acbc37b9/ouroboros-consensus-cardano/src/shelley/Ouroboros/Consensus/Shelley/Node.hs#L75-L76

Any intuition about whether this is worth to pursue/assess? Currently, I think I will just mention this in a comment as a potential opportunity.

Reply:
I think this is inconsequential, resolving.